node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/Activator.java

```diff
@@ -53,41 +57,69 @@ public void activate(Collection<HostSpec> hosts, NestedTransaction transaction,
+        List<Node> reservedToActivate = updatePortsFrom(hosts, retainHostsInList(hostnames, reserved));
+        List<Node> oldActive = applicationNodes.state(Node.State.active).asList(); // All nodes active now
+        List<Node> continuedActive = retainHostsInList(hostnames, oldActive);
+        List<Node> newActive = updateFrom(hosts, continuedActive); // All nodes that will be active when this is committed
+        newActive.addAll(reservedToActivate);
+        if ( ! containsAll(hostnames, newActive))
             throw new IllegalArgumentException("Activation of " + application + " failed. " +
                                                "Could not find all requested hosts." +
                                                "\nRequested: " + hosts +
                                                "\nReserved: " + toHostNames(reserved) +
-                                               "\nActive: " + toHostNames(active) +
+                                               "\nActive: " + toHostNames(oldActive) +
                                                "\nThis might happen if the time from reserving host to activation takes " +
                                                "longer time than reservation expiry (the hosts will then no longer be reserved)");
 
         validateParentHosts(application, allNodes, reservedToActivate);
 
-        List<Node> activeToRemove = removeHostsFromList(hostnames, active);
-        activeToRemove = activeToRemove.stream().map(Node::unretire).collect(Collectors.toList()); // only active nodes can be retired
-        nodeRepository.deactivate(activeToRemove, transaction, lock);
-        nodeRepository.activate(updateFrom(hosts, continuedActive), transaction); // update active with any changes
-        nodeRepository.activate(updatePortsFrom(hosts, reservedToActivate), transaction);
+        List<Node> activeToRemove = removeHostsFromList(hostnames, oldActive);
+        activeToRemove = activeToRemove.stream().map(Node::unretire).collect(Collectors.toList()); // only active nodes can be retired. TODO: Move this line to deactivate
+        nodeRepository.deactivate(activeToRemove, transaction);
+        nodeRepository.activate(newActive, transaction.nested()); // activate also continued active to update node state
+
+        rememberResourceChange(transaction, generation,
+                               NodeList.copyOf(oldActive).not().retired(),
+                               NodeList.copyOf(newActive).not().retired());
         unreserveParentsOf(reservedToActivate);
     }
 
+    private void rememberResourceChange(ApplicationTransaction transaction, long generation,
+                                        NodeList oldNodes, NodeList newNodes) {
+        if (nodeRepository.applications().get(transaction.application()).isEmpty()) return; // infrastructure app, hopefully
+        Application application = nodeRepository.applications().get(transaction.application()).get();
```

Comment:
Read this once from ZK?

Reply:
This is reading it once?

Reply:
It's reading it in the if statement above as well?

Reply:
That is true. Changed.